### PR TITLE
8295290: Add Windows ARM64 ABI support to the Foreign Function & Memory API

### PR DESCRIPTION
**Pull request by swesonga**
There are 2 primary differences between the Windows ARM64 ABI and the macOS/Linux ARM64 ABI: variadic floating point arguments are passed in general purpose registers on Windows (instead of the vector registers). In addition to this, up to 64 bytes of a struct being passed to a variadic function can be placed in general purpose registers. This happens regardless of the type of struct (HFA or other generic struct). This means that a struct can be split across registers and the stack when invoking a variadic function.

This change introduces tests that compute the sum of the fields of structs containing 1-4 ints, floats, and doubles to verify that each field is correctly assigned a register or stack location when invoking a variadic function (both when the struct can be passed entirely in registers as well as when the struct spills onto the stack).

For details about the Foreign Function & Memory API, see JEP 434 defined at https://openjdk.org/jeps/434

The Windows ARM64 ABI conventions are documented at https://learn.microsoft.com/en-us/cpp/build/arm64-windows-abi-conventions

This work builds on @lewurm / Bernhard's branch at https://github.com/lewurm/openjdk/commits/foreign-windows-aarch64

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Change must be properly reviewed (1 review required, with at least 1 [Committer](https://openjdk.org/bylaws#committer))

### Issue
 * [JDK-8295290](https://bugs.openjdk.org/browse/JDK-8295290): Add Windows ARM64 ABI support to the Foreign Function & Memory API

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/panama-foreign pull/754/head:pull/754` \
`$ git checkout pull/754`

Update a local copy of the PR: \
`$ git checkout pull/754` \
`$ git pull https://git.openjdk.org/panama-foreign pull/754/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 754`

View PR using the GUI difftool: \
`$ git pr show -t 754`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/panama-foreign/pull/754.diff">https://git.openjdk.org/panama-foreign/pull/754.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64VaList.java
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64VaList.java:147]: Calls to `toSessionImpl` can be replaced with simple cast
```suggestion
        ((MemorySessionImpl) segment.session()).checkValidState();
```

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64VaList.java
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64VaList.java:174]:
```suggestion
        ((MemorySessionImpl) segment.session()).checkValidState();
```

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64VaList.java
Comment (JornVernee) [inline src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/windows/WindowsAArch64VaList.java:189]:
```suggestion
            ((MemorySessionImpl) segment.session()).checkValidState();
```

**Review by JornVernee — COMMENTED** [review]
This looks like a really good start, but there are few things to be done still, I think.

Note also that I'm not able to test this on Windows/AArch64, though I've run `jdk_foreign` tests on all the other supported platforms. Some tests having sampling applied, so it is important to run tests with `-Dgenerator.sample.factor=1`. When running tests through `make` this can be done by using `JTREG="JAVA_OPTIONS=-Dgenerator.sample.factor=1"`.

We also have the test `java/foreign/TestMatrix.java` which runs several different configurations. It's a manual test that has to be run through `jtreg` directly. A comment at the top of that file contains instructions to do that. It's also a good idea to run that test (and it isn't run when using `make`). This test also fully runs all the sampled tests at, except for TestVarArgs (we should probably add a cases for that...).

Lastly, eventually this patch will have to be brought over to the mainline JDK repo as well. I suggest you take charge of that as well, since you can do the testing on Windows/AArch64. It might also be easier to straight up integrate things into the mainline JDK repo after the changes for the JDK 20 go into that repo [1], [2] (will likely be somewhere in the next 2 weeks). After that the code is again merged into this repo, and both repos should contain the same code any ways. (There might be some changes needed to the code as well, due to naming changes).

But, we can get started on the review here already.

[1]: https://github.com/openjdk/jdk/pull/10872
[2]: https://github.com/openjdk/jdk/pull/11019

**Comment by JornVernee** [comment]
@swesonga I applied some fixes to test/jdk/java/foreign/TestMatrix.java [1] that were merged into this repo today. You might want to merge the latest `foreign-memaccess+abi` branch into this PR branch to get those changes as well.

[1]: https://github.com/openjdk/jdk/pull/11398

**Review by JornVernee — COMMENTED** [review]
Few more comments inline.

FWIW, I think CallArranger can be cleaned up further, and I've been playing with a patch [1]. But I'll file a followup PR for that.

[1]: https://github.com/openjdk/panama-foreign/compare/pull/754/head...JornVernee:panama-foreign:Refector_CallArranger

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java
Comment (swesonga) [reply src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:287]: Yes, Windows on ARM64 requires these floats to be in integer registers when calling variadic functions.

File: src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java
Comment (swesonga) [reply src/java.base/share/classes/jdk/internal/foreign/abi/aarch64/CallArranger.java:263]: Yes, that is correct. An HFA with 4 floats that is the last non-variadic argument after 6 other non-variadic arguments will have half of it passed in x7 and half at [sp].

File: test/jdk/java/foreign/valist/VaListTest.java
Comment (swesonga) [reply test/jdk/java/foreign/valist/VaListTest.java]: Deleted this file in latest merge (deleted by https://github.com/openjdk/panama-foreign/commit/4617dee95b649a5eb7fbd8a83d4e2c8af1250116)

**Comment by swesonga** [comment]
> Few more comments inline.
> 
> FWIW, I think CallArranger can be cleaned up further, and I've been playing with a patch [1](https://github.com/openjdk/panama-foreign/compare/pull/754/head...JornVernee:panama-foreign:Refector_CallArranger). But I'll file a followup PR for that.

@JornVernee just to clarify, do you expect more refactoring of the CallArranger in this PR or will that be subsumed by your patch?

**Comment by JornVernee** [comment]
@swesonga No, I think this PR is good. I'll file a followup PR myself for the other refactoring I did.

**Review by JornVernee — APPROVED** [review]

**Comment by JornVernee** [comment]
@swesonga Hey, just a heads up. It is about that time again where we are looking to put together a PR for the JEP to go into mainline.

Since we do not support Windows/AArch64 (e.g. we can not test it), we can not be in charge of a PR that brings this port over to mainline. So, I think it's up to you or another maintainer of Windows/AArch64.

I suggest taking the following two commits:
1. https://github.com/openjdk/panama-foreign/commit/d379ca1ca32926591fbf8a4afdae1a4a52b3cb63
2. https://github.com/openjdk/panama-foreign/commit/08225e4f5e00837e974098a933f27267f3c0dbc5

And making a PR against openjdk/jdk out of those (since those three are all related to this port). I can approve it again there, since I've already reviewed the changes before.

**Comment by swesonga** [comment]
> @swesonga Hey, just a heads up. It is about that time again where we are looking to put together a PR for the JEP to go into mainline.
> 
> Since we do not support Windows/AArch64 (e.g. we can not test it), we can not be in charge of a PR that brings this port over to mainline. So, I think it's up to you or another maintainer of Windows/AArch64.
> 
> I suggest taking the following two commits:
> 
> 1. [d379ca1](https://github.com/openjdk/panama-foreign/commit/d379ca1ca32926591fbf8a4afdae1a4a52b3cb63)
> 2. [08225e4](https://github.com/openjdk/panama-foreign/commit/08225e4f5e00837e974098a933f27267f3c0dbc5)
> 
> And making a PR against openjdk/jdk out of those (since those three are all related to this port). I can approve it again there, since I've already reviewed the changes before.

@JornVernee I've opened https://github.com/openjdk/jdk/pull/12773

**Pull request closed without merging**
